ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala

```diff
@@ -55,22 +64,134 @@ object EndpointsCompanion {
+    )(implicit
+        mf: Monad[Future]
+    ): EitherT[Future, Unauthorized, B] = {
+      import scalaz.syntax.std.option._
+      import scalaz.syntax.std.either._
+      import spray.json._
+      for {
+        jsValue <- either(
+          Try(jwt.payload.parseJson).toEither.disjunction
+            .leftMap(it => Unauthorized(it.getMessage))
+        )
+        decodedToken <- either(
+          readStandardTokenPayload(
+            jsValue
+          ) \/> Unauthorized("Invalid token supplied")
+        )
+
+        userId <- either(
+          (decodedToken.applicationId \/> Unauthorized(
+            "This user token contains no applicationId/userId"
+          )).flatMap(UserId.fromString(_).left.map(Unauthorized).disjunction)
+        )
+        rights <- EitherT.rightT(listUserRights(userId))
+
+        actAs = rights.collect { case CanActAs(party) =>
+          party
+        }
+        readAs = rights.collect { case CanReadAs(party) =>
+          party
+        }
+        ledgerId <- EitherT.rightT(getLedgerId())
+        res <- either(fromUser(userId, actAs.toList, readAs.toList, ledgerId))
+      } yield res
+    }
+
+    private[http] implicit def jwtWritePayloadFromUserToken(implicit
+        mf: Monad[Future]
+    ): CreateFromUserToken[JwtWritePayload] =
+      (
+          jwt,
+          listUserRights,
+          getLedgerId,
+      ) =>
+        transformUserTokenTo(jwt, listUserRights, getLedgerId)((userId, actAs, readAs, ledgerId) =>
+          for {
+            actAsNonEmpty <-
+              if (actAs.isEmpty)
+                -\/ apply Unauthorized(
+                  "ActAs list of user was empty, this is an invalid state for converting it to a JwtWritePayload"
+                )
+              else \/-(NonEmptyList(actAs.head: String, actAs.tail: _*))
+          } yield JwtWritePayload(
+            lar.LedgerId(ledgerId),
+            lar.ApplicationId(userId),
+            lar.Party.subst(actAsNonEmpty),
+            lar.Party.subst(readAs),
+          )
+        )
+
+    private[http] implicit def jwtPayloadLedgerIdOnlyFromUserToken(implicit
+        mf: Monad[Future]
+    ): CreateFromUserToken[JwtPayloadLedgerIdOnly] =
+      (_, _, getLedgerId: () => Future[String]) =>
+        EitherT
+          .rightT(getLedgerId())
+          .map(ledgerId => JwtPayloadLedgerIdOnly(lar.LedgerId(ledgerId)))
+
+    private[http] implicit def jwtPayloadFromUserToken(implicit
+        mf: Monad[Future]
+    ): CreateFromUserToken[JwtPayload] =
+      (
+          jwt,
+          listUserRights,
+          getLedgerId,
+      ) =>
+        transformUserTokenTo(jwt, listUserRights, getLedgerId)((userId, actAs, readAs, ledgerId) =>
+          \/ fromEither JwtPayload(
+            lar.LedgerId(ledgerId),
+            lar.ApplicationId(userId),
+            actAs = lar.Party.subst(actAs),
+            readAs = lar.Party.subst(readAs),
+          ).toRight(Unauthorized("Something went wrong and IDK what I can write here :D"))
```

Comment:
if you don’t want to address this in this PR, add a TODO or fixme with an issue reference.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/12075/commits/0f248d24fdbc9cf10de9e48a671c0f2d03bf7319